cmd/cmd.go

```diff
@@ -773,59 +775,87 @@ func CopyHandler(cmd *cobra.Command, args []string) error {
 }
 
 func PullHandler(cmd *cobra.Command, args []string) error {
-	insecure, err := cmd.Flags().GetBool("insecure")
-	if err != nil {
-		return err
-	}
+	name := args[0] // cobra should be cofigured to always pass 1 arg
 
 	client, err := api.ClientFromEnvironment()
 	if err != nil {
 		return err
 	}
 
-	p := progress.NewProgress(os.Stderr)
-	defer p.Stop()
-
-	bars := make(map[string]*progress.Bar)
-
-	var status string
-	var spinner *progress.Spinner
-
-	fn := func(resp api.ProgressResponse) error {
-		if resp.Digest != "" {
-			if spinner != nil {
-				spinner.Stop()
-			}
-
-			bar, ok := bars[resp.Digest]
-			if !ok {
-				bar = progress.NewBar(fmt.Sprintf("pulling %s...", resp.Digest[7:19]), resp.Total, resp.Completed)
-				bars[resp.Digest] = bar
-				p.Add(resp.Digest, bar)
-			}
+	ctx, cancel := context.WithCancel(cmd.Context())
+	defer cancel()
 
-			bar.Set(resp.Completed)
-		} else if status != resp.Status {
-			if spinner != nil {
-				spinner.Stop()
+	errc := make(chan error, 1)
+	var mu sync.Mutex
+	progressLocked := make(map[string][2]int64) // digest -> [completed, total]
+	go func() {
+		p := &api.PullRequest{Name: name}
+		errc <- client.Pull(ctx, p, func(up api.ProgressResponse) error {
+			if up.Digest == "" && up.Status != "" {
+				// A status with no digest is a terminal
+				// status. Give up and return an error.
+				//
+				// First, strip any "error:" prefix so it does
+				// not stutter with Cobra's "Error: " prefix.
+				//
+				// Our future client will handle the stream
+				// updates and errors properly. This works for
+				// now though.
+				status := strings.TrimPrefix(up.Status, "error: ")
+				return errors.New(status)
 			}
+			mu.Lock()
+			progressLocked[up.Digest] = [2]int64{up.Completed, up.Total}
+			mu.Unlock()
+			return nil
+		})
+	}()
 
-			status = resp.Status
-			spinner = progress.NewSpinner(status)
-			p.Add(status, spinner)
+	t := time.NewTicker(100 * time.Millisecond) // "unstarted"
+	defer t.Stop()
+
+	writeHideCursor(os.Stderr)
+	defer writeShowCursor(os.Stderr)
+	writeSavePos(os.Stderr)
+
+	progress := make(map[string][2]int64)
+	flushProgress := func() {
+		mu.Lock()
+		maps.Copy(progress, progressLocked)
+		mu.Unlock()
+		var completed, total int64
+		for _, v := range progress {
+			completed += v[0]
+			total += v[1]
+		}
+		if total > 0 {
+			writeRestorePos(os.Stderr)
+			fmt.Printf("Pulling %s %d/%d %0.1f%%\n",
```

Comment:
```suggestion
			fmt.Printf("Downloading %s %d/%d %0.1f%%\n",
```

For large files, `completed` and `total` are really big

```
Pulling gemma3:27b 17396936941/17396936941 100.0%
```

do we want to show the size in human format?

Reply:
The answer is obvious. It was a minor oversight before pushing. I'm on it.